<div style="border:solid Chocolate 2px; padding: 40px">

<b> Николай, привет!👋</b>

Меня зовут Алексей Гриб, и я буду ревьюером твоего проекта. 

Сразу хочу предложить в дальнейшем общаться на "ты" - надеюсь, так будет комфортнее:) Но если это неудобно, обязательно дай знать, и мы придумаем что-нибудь ещё!
    
Цель ревью - не искать ошибки в твоём проекте, а помочь тебе сделать твою работу ещё лучше, устранив недочёты и приблизив её к реальным задачам аналитика. Поэтому не расстраивайся, если что-то не получилось с первого раза - это нормально, и это поможет тебе вырасти!
    
Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> похвала, рекомендации «со звёздочкой», полезные лайфхаки, которые сделают и без того красивое решение ещё более элегантным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> некритичные ошибки или развивающие рекомендации на будущее. 
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
Критичные ошибки, которые обязательно нужно исправить.
</div>

Я не смогу принять проект, если в нём будет хотя бы одна критичная ошибка или несколько некритичных ошибок - тогда проект нужно будет немного доработать. Но это нестрашно - я обязательно дам тебе подсказку или укажу верное направление.
    
Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 
    
Ты также можешь задавать свои вопросы, реагировать на мои комментарии, делать пометки и пояснения - полная творческая свобода! Но маленькая просьба - пускай они будут отличаться от моих комментариев, это поможет избежать путаницы в нашем общении:)
Например, вот так:
    
<br/>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*твой текст*
</div>
    
Давай посмотрим на твой проект!

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта." data-toc-modified-id="Описание-проекта.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта.</a></span></li><li><span><a href="#Описание-данных." data-toc-modified-id="Описание-данных.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных.</a></span></li><li><span><a href="#Задания-по-проекту." data-toc-modified-id="Задания-по-проекту.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задания по проекту.</a></span></li><li><span><a href="#Импорт-библиотек,-подключение-к-БД,-первичный-анализ-данных." data-toc-modified-id="Импорт-библиотек,-подключение-к-БД,-первичный-анализ-данных.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Импорт библиотек, подключение к БД, первичный анализ данных.</a></span></li><li><span><a href="#Выполнение-заданий-по-проекту." data-toc-modified-id="Выполнение-заданий-по-проекту.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выполнение заданий по проекту.</a></span><ul class="toc-item"><li><span><a href="#Посчитаем,-сколько-книг-вышло-после-1-января-2000-года." data-toc-modified-id="Посчитаем,-сколько-книг-вышло-после-1-января-2000-года.-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Посчитаем, сколько книг вышло после 1 января 2000 года.</a></span></li><li><span><a href="#Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку." data-toc-modified-id="Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку.-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Для каждой книги посчитаем количество обзоров и среднюю оценку.</a></span></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц." data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц.-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Издательство, которое выпустило наибольшее число книг толще 50 страниц.</a></span></li><li><span><a href="#Определим-автора-с-самой-высокой-средней-оценкой-книг-—-будем-учитывать-только-книги-с-50-и-более-оценками." data-toc-modified-id="Определим-автора-с-самой-высокой-средней-оценкой-книг-—-будем-учитывать-только-книги-с-50-и-более-оценками.-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Определим автора с самой высокой средней оценкой книг — будем учитывать только книги с 50 и более оценками.</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li><li><span><a href="#Вывод." data-toc-modified-id="Вывод.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод.</a></span></li></ul></div>

# Проект по SQL.

## Описание проекта.

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных.

Таблица `books` - Содержит данные о книгах:

- `book_id` — идентификатор книги
- `author_id` — идентификатор автора
- `title` — название книги
- `num_pages` — количество страниц
- `publication_date` — дата публикации книги
- `publisher_id` — идентификатор издателя

Таблица `authors` - Содержит данные об авторах:

- `author_id` — идентификатор автора
- `author` — имя автора

Таблица `publishers` - Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства
- `publisher` — название издательства

Таблица `ratings` - Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки
- `book_id` — идентификатор книги
- `username` — имя пользователя, оставившего оценку
- `rating` — оценка книги

Таблица `reviews` - Содержит данные о пользовательских обзорах:

- `review_id` — идентификатор обзора
- `book_id` — идентификатор книги
- `username` — имя автора обзора
- `text` — текст обзора

## Задания по проекту.

- Посчитать, сколько книг вышло после 1 января 2000 года
- Для каждой книги посчитать количество обзоров и среднюю оценку
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключить из анализа брошюры
- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Прекрасное вступление!
    
В нём есть всё, что необходимо, чтобы понять суть проекта с первых строк отчёта!
    
Спасибо, что есть оглавление в плагине toc и в виде гиперссылок - по проекту удобно перемещаться:)

## Импорт библиотек, подключение к БД, первичный анализ данных.

In [1]:
# Импортируем библиотеки

import pandas as pd
from sqlalchemy import create_engine

In [2]:
# Подключаемся к БД

db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# Сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Библиотеки импортировали, коннектор сделали - отлично!
    
Спасибо, что перезапускаешь ядро проекта перед отправкой на ревью - это важный шаг, который поможет избежать простых и досадных ошибок при воспроизведении решнеия:)
</div>

In [3]:
# Исследуем таблицы, выведем их первые строки

for i in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    print('Таблица ', i)
    display(pd.io.sql.read_sql(f'''SELECT * FROM {i} LIMIT 5''', con = engine))
    print()
    print()

Таблица  books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268



Таблица  authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd



Таблица  publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company



Таблица  ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2



Таблица  reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Первичный осмотр данных сделали - классно, что процесс организован через цикл, отличная работа!
</div>

## Выполнение заданий по проекту.

### Посчитаем, сколько книг вышло после 1 января 2000 года.

In [4]:
query = '''
SELECT
    COUNT(book_id)
FROM
    books
WHERE
    publication_date > '2000-01-01'
'''

pd.io.sql.read_sql(query, con = engine)

count
0    819

После 1-го января 2000 года вышло 819 книг.

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Правильно!
    
Отмечу, что условие фильтрации `publication_date >= '2000-01-01' ` также было бы правильным, хоть и привело бы к другому результату - в реальных проектах о том, включается ли дата в диапазон или нет, всегда стоит уточнять заранее:)

</div>

### Для каждой книги посчитаем количество обзоров и среднюю оценку.

In [5]:
query1 = '''
SELECT
    books.book_id,
    books.title,
    COUNT(DISTINCT review_id) as count_reviews,
    AVG(rating) as avg_rating
FROM 
    books
FULL JOIN reviews ON  reviews.book_id = books.book_id
FULL JOIN ratings ON  ratings.book_id = books.book_id
GROUP BY
    books.book_id
ORDER BY
    count_reviews DESC,
    avg_rating DESC
LIMIT 10
'''
pd.io.sql.read_sql(query1, con = engine)

book_id                                              title  count_reviews  \
0      948                            Twilight (Twilight  #1)              7   
1      302  Harry Potter and the Prisoner of Azkaban (Harr...              6   
2      299  Harry Potter and the Chamber of Secrets (Harry...              6   
3      656                                     The Book Thief              6   
4      734                                   The Glass Castle              6   
5      497                          Outlander (Outlander  #1)              6   
6      750                The Hobbit  or There and Back Again              6   
7      695  The Curious Incident of the Dog in the Night-Time              6   
8      779  The Lightning Thief (Percy Jackson and the Oly...              6   
9      963                                Water for Elephants              6   

   avg_rating  
0    3.662500  
1    4.414634  
2    4.287500  
3    4.264151  
4    4.206897  
5    4.125000  
6    4.125000  
7    4.081081  
8    4.080645  
9    3.977273

Больше всего обзоров на книгу `Twilight (Twilight #1)` - 7 обзоров.

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Тут всё верно! Молодец, что посчитал метрики для всех книг, даже для тех, у которых нет отзывов!

</div>

### Издательство, которое выпустило наибольшее число книг толще 50 страниц.

In [6]:
query2 = '''
SELECT
    publisher,
    COUNT(book_id) as count_book
FROM
    publishers
FULL JOIN books ON  books.publisher_id = publishers.publisher_id 
WHERE 
    num_pages > 50
GROUP BY
    publisher
ORDER BY
    count_book DESC
LIMIT 1
'''
pd.io.sql.read_sql(query2, con = engine)

publisher  count_book
0  Penguin Books          42

Наибольшее число книг выпустило издательство `Penguin Books`.

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Издательство определили верно! Спасибо, что дополнительно вывел количество книг издательства!

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
В задании сказано, что нужно найти одно издательство, поэтому было бы логичнее оставить в выовде только одну строку:)
        </div>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Поправил.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено!

</div>

### Определим автора с самой высокой средней оценкой книг — будем учитывать только книги с 50 и более оценками.

In [7]:
query3 = '''
WITH data as (
SELECT
   a.author,
   r.book_id,
   COUNT(rating) as count_rating,
   AVG(rating) as avg_rating
FROM
   ratings r
LEFT JOIN books b ON r.book_id = b.book_id
LEFT JOIN authors a ON b.author_id = a.author_id
GROUP BY
    a.author,
    r.book_id
HAVING
   COUNT(rating) > 50
   )

SELECT
    author,
    AVG(avg_rating) as avg_rating
FROM
    data
GROUP BY
    author
ORDER BY
    avg_rating DESC
    LIMIT 1
'''
pd.io.sql.read_sql(query3, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

Самая высокая средняя оценка книг у авторов `J.K. Rowling/Mary GrandPré` - 4.28.

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Автор с самым высоким средним рейтингом найден верно - молодец, это одно из самых сложных заданий проекта, и ты сходу решил его правильно! 
    
Спасибо, что дополнитльно вывел средний рейтинг автора!

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
В задании сказано, что нужно найти одного автора, поэтому было бы логичнее оставить в выовде только одну строку:)
        </div>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Поправил.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено!

</div>

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [8]:
query4 = '''
        SELECT ROUND(AVG(cnt_reviews)) AS avg_reviews
        FROM
        (
        SELECT COUNT(review_id) AS cnt_reviews
        FROM
        (SELECT username 
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50) AS sub
        LEFT JOIN reviews ON sub.username = reviews.username
        GROUP BY sub.username) AS sub_2
        '''
pd.io.sql.read_sql(query4, con = engine)

avg_reviews
0         24.0

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Правильно!
    
Среднее количество обзоров от пользователей, которые поставили больше 50 оценок, также посчитано верно! Спасибо, что округлил результат до целого!

</div>

## Вывод.

Подключились к базе данных PostgreSQL, ознакомились с содержимым 5 таблиц: `books`, `authors`, `publishers`, `ratings`, `reviews`.

Анализ данных показал, что:

- после 1 января 2000 года вышло 819 книг
- На книгу `Twilight (Twilight #1)` вышло максимальное количество обзоров - 7, рейтинг книги - 3.7
- издательство, выпустившее наибольшее число книг толще 50 страниц: `Penguin Books`, они выпустили 42 книги
- авторы с самой высокой средней оценкой книг - `J.K. Rowling/Mary GrandPré` с рейтингом 4.28
- среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Отличный завершающий вывод!

</div>

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту.

Николай, твой проект вышел на высоте - все задания решены верно!

Мне нравится твой аналитический подход к выполнению задачи, ты соблюдаешь структуру работы, выполняешь её последовательно - это очень хорошо! Шаги проекта выполнены по порядку согласно плану проекта, нет смысловых и структурных ям. Важно, что не забываешь про выводы.
    
SQL запросы написаны хорошо и оптимально, они возвращают верный результат, в них нет избыточных действий - это здорово, кажется, ты отлично разобрался в теме!
    
Также в работе есть небольшие рекомендации по улучшению описательной части - это мелочь по оформлению твоих ответов, которая не займёт много времени, но сделает твой проект стилистически более совершенным.
    
Проект можно было бы принять с первого раза - его исполнение практически безупречно, а жёлтые комментарии крайне несущественные. Однако на случай, если у тебя есть какие-то вопросы по проекту/SQL/домену или всему, чему угодно, верну проект для Q&A. Ну и если ты учтёшь рекомендации по оформлению ответов - будет классно:) 
    
После того, как отправишь мне проект снова - приму его вне очереди:)

    
Жду твой проект для завершения работы над ним. До встречи:)

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Замечания учел. Вопросов по проекту SQL не имею)
Спасибо!)
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту v.2.

Николай, спасибо за оперативные правки - проект принимаю:)
    
Желаю успехов в завершении обучения и дальнейшем трудоустройстве:)